interfaces/builtin/cups_control.go

```diff
@@ -19,30 +19,159 @@
+
+# Allow daemon to send signals to its snap_daemon processes
+capability kill,
+
+# Allow daemon to manage snap_daemon files and directories
+capability fsetid,
+`
+
+const cupsControlConnectedSlotAppArmor = `
+# Allow daemon to send notifications to connected snaps
+dbus (send)
+    bus=system
+    path=/org/cups/cupsd/Notifier
+    interface=org.cups.cupsd.Notifier
+    peer=(name=org.freedesktop.DBus,label=###PLUG_SECURITY_TAGS###),
 `
 
 const cupsControlConnectedPlugAppArmor = `
-# Description: Can access cups control socket. This is restricted because it provides
-# privileged access to configure printing.
+# Allow communicating with the cups server for printing and configuration.
 
 #include <abstractions/cups-client>
-/run/cups/printcap r,
+/{,var/}run/cups/printcap r,
+
+# Allow receiving all DBus signal notifications from the daemon (see
+# notifier/dbus.c in cups sources)
+dbus (receive)
+    bus=system
+    path=/org/cups/cupsd/Notifier
+    interface=org.cups.cupsd.Notifier
+    peer=(name=org.freedesktop.DBus,label=###SLOT_SECURITY_TAGS###),
 `
 
+type cupsControlInterface struct {
+	commonInterface
+}
+
+func (iface *cupsControlInterface) AppArmorPermanentSlot(spec *apparmor.Specification, slot *snap.SlotInfo) error {
+	// Only apply slot snippet when running as application snap on classic
+	// since the slot side may be from the classic OS or snap.
+	if !implicitSystemPermanentSlot(slot) {
+		spec.AddSnippet(cupsControlPermanentSlotAppArmor)
+	}
+	return nil
+}
+
+func (iface *cupsControlInterface) AppArmorConnectedSlot(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	// Only apply slot snippet when running as application snap on classic
+	// since the slot side may be from the classic OS or snap.
+	if !implicitSystemConnectedSlot(slot) {
+		old := "###PLUG_SECURITY_TAGS###"
+		new := plugAppLabelExpr(plug)
+		snippet := strings.Replace(cupsControlConnectedSlotAppArmor, old, new, -1)
+		spec.AddSnippet(snippet)
+	}
+	return nil
+}
+
+func (iface *cupsControlInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	old := "###SLOT_SECURITY_TAGS###"
+	var new string
+	// If we're running on classic, cups may be installed either as a snap
+	// or as part of the classic OS. If it is part of the classic OS, it
+	// will not have a security label like it would when installed as a
+	// snap.
+	if implicitSystemConnectedSlot(slot) {
+		// cupsd from the OS may be confined or unconfined. Newer
+		// releases may use the 'cupsd' label instead of the old
+		// path-based label.
+		new = "\"{unconfined,/usr/sbin/cupsd,cupsd}\""
+	} else {
+		new = slotAppLabelExpr(slot)
+	}
+
+	// implement 'implicitOnCore: false/implicitOnClassic: true' by only
+	// applying the snippet if the slot is an app or we are on classic
+	if slot.Snap().Type() == snap.TypeApp || release.OnClassic {
```

Comment:
@anonymouse64 - when I was developing the test cases and looked at avahi-control/observe, I discovered that we didn't have complete coverage and there was nothing in AppArmorConnectedPlug() that was making sure 'implicitOnCore: false/implicitOnClassic: true' was honored. Eg, if I comment out this conditional, then `// core to consumer on core is empty for ConnectedPlug` fails in `TestAppArmorSpecCore`, which is completely expected since we are calling AddConnectedPlug() directly.

Especially since the concept of "implicit on classic or app installed on core and classic" is relatively new (as mentioned before, it is only avahi* and cups-control at this point), I wanted to err on the side of caution and have unit tests that explicitly showed it was working properly. If you'd prefer that I remove this conditional and we just rely on implicit* working as they should, I can see that point and can adjust.

Reply:
This is fine, you can leave this in, I just wanted to make sure I wasn't missing something and that this logic was just extra caution.

Reply:
UPDATE: there are actually several interfaces besides avahi* and cups that have this base decl/implicit* combination, so I added a new test for these.